lib-injection/requirements.rb

```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/env ruby
+
+require 'json'
+
+def requirements
+  reqs = {
+    '$schema' => 'https://raw.githubusercontent.com/DataDog/auto_inject/refs/heads/main/preload_go/cmd/library_requirements_tester/testdata/requirements_schema.json',
+    'version' => 1,
+    'native_deps' => {
+      'glibc' => [
+        {
+          'arch' => 'x64',
+          'supported' => true,
+          'min' => '2.27',
+          'description' => 'libffi needs memfd_create',
+        },
+        {
+          'arch' => 'arm64',
+          'supported' => true,
+          'min' => '2.27',
+          'description' => 'libffi needs memfd_create',
+        },
+      ],
+      'musl' => [
+        {
+          'arch' => 'x64',
+          'supported' => false,
+          'description' => 'no musl build',
+        },
+        {
+          'arch' => 'arm64',
+          'supported' => false,
+          'description' => 'no musl build',
+        },
+      ],
+    },
+    'deny' => [],
+  }
+
+  # gem command such as install aren't quite usuful to inject into
+  # plus there are issues when it ends up calling gcc
+  reqs['deny'] << {
+    'id' => 'gem',
+    'description' => 'Ignore the rubygems CLI',
+    'os' => nil,
+    'cmds' => [
+      '**/ruby'
+    ],
+    'args' => [{ 'args' => ['**/gem'], 'position' => 1 }],
+    'envars' => nil,
+  }
+
+  # `bundle exec` is the only command we want to inject into.
+  # there is no `allow` overriding `deny` so we're left to exclude all of the
+  # possible others.
+  %w[bundle bundler].each do |exe|
+    %w[
+      install
+      update
+      cache
+      config
+      help
+      add
+      binstubs
+      check
+      show
+      outdated
+      console
+      open
+      lock
+      viz
+      init
+      gem
+      platform
+      clean
+      doctor
+      remove
+      plugin
+      version
```

Comment:
@andrewlock This is supremely annoying, basically we want "deny everything but `bundle exec`".

This could be achieved with an `allow:` key having higher precedence than `deny:`, defaulting to `allow: []`, thus:

- anything allowed would not go through `deny:`
- is backwards compatible

```
  "allow": [
    {
      "id": "bundle_exec",
      "description": "Allow bundle exec",
      "os": null,
      "cmds": [
        "**/ruby"
      ],
      "args": [
        {
          "args": [
            "**/bundle",
            "exec"
          ],
          "position": 1
        }
      ],
      "envars": null
    }
  ],
  "deny": [
    {
      "id": "bundle_not_exec",
      "description": "Deny everything but bundle exec",
      "os": null,
      "cmds": [
        "**/ruby"
      ],
      "args": [
        {
          "args": [
            "**/bundle"
          ],
          "position": 1
        }
      ],
      "envars": null
    }
  ],
```

Reply:
Not only annoying, but with new command added we will chase the edge all the time. What you've suggested is way better (and safer)

Reply:
I totally agree with this.